test/unittests/t_util.cc

```diff
@@ -215,3 +314,855 @@ TEST(T_Util, IsHttpUrl) {
+  EXPECT_EQ(pstr_fake_path, GetParentPath(pstr_fake_path));
+}
+
+TEST_F(T_Util, GetFileName) {
+  EXPECT_EQ(empty, GetFileName(path_with_slash));
+  EXPECT_EQ("path", GetFileName(path_without_slash));
+  EXPECT_EQ(fake_path, GetFileName(fake_path));
+
+  EXPECT_EQ(NameString(empty), GetFileName(PathString(path_with_slash)));
+  EXPECT_EQ(NameString(NameString("path")),
+      GetFileName(PathString(path_without_slash)));
+  EXPECT_EQ(NameString(fake_path), GetFileName(PathString(fake_path)));
+}
+
+TEST_F(T_Util, CreateFile) {
+  ASSERT_DEATH(CreateFile("myfakepath/otherfakepath.txt", 0777), ".*");
+  string filename = string(sandbox) + "/createfile.txt";
+  CreateFile(filename, 0600);
+  FILE* myfile = fopen(filename.c_str(), "w");
+  EXPECT_NE(static_cast<FILE*>(NULL), myfile);
+  fclose(myfile);
+}
+
+TEST_F(T_Util, MakeSocket) {
+  int socket_fd1;
+  int socket_fd2;
+
+  ASSERT_DEATH(MakeSocket(long_path, 0600), ".*");
+  EXPECT_NE(-1, socket_fd1 = MakeSocket(string(socket_address), 0777));
+  // the second time it should work as well (non socket-alrady-in-use error)
+  EXPECT_NE(-1, socket_fd2 = MakeSocket(string(socket_address), 0777));
+  close(socket_fd1);
+  close(socket_fd2);
+}
+
+TEST_F(T_Util, ConnectSocket) {
+  int server_fd = MakeSocket(string(socket_address), 0777);
+  listen(server_fd, 1);
+  int client_fd = ConnectSocket(string(socket_address));
+
+  ASSERT_DEATH(ConnectSocket(string(long_path)), ".*");
+  ASSERT_EQ(-1, ConnectSocket(string(sandbox) + "/fake_socket"));
+  ASSERT_NE(-1, client_fd);
+  close(client_fd);
+  close(server_fd);
+}
+
+TEST_F(T_Util, MakePipe) {
+  int fd[2];
+  void *buffer_output = calloc(100, sizeof(char));
+  MakePipe(fd);
+  write(fd[1], to_write, string(to_write).length());
+  read(fd[0], buffer_output, string(to_write).length());
+
+  EXPECT_STREQ(to_write, static_cast<const char*>(buffer_output));
+  ASSERT_DEATH(MakePipe(static_cast<int*>(NULL)), ".*");
+  free(buffer_output);
+  ClosePipe(fd);
+}
+
+TEST_F(T_Util, WritePipe) {
+  int fd[2];
+  void *buffer_output = calloc(20, sizeof(char));
+  MakePipe(fd);
+  WritePipe(fd[1], to_write, string(to_write).length());
+  read(fd[0], buffer_output, string(to_write).length());
+
+  EXPECT_STREQ(to_write, static_cast<const char*>(buffer_output));
+  ASSERT_DEATH(WritePipe(-1, to_write, string(to_write).length()),
+      ".*");
+  free(buffer_output);
+  ClosePipe(fd);
+}
+
+TEST_F(T_Util, ReadPipe) {
+  int fd[2];
+  void *buffer_output = calloc(20, sizeof(char));
+  MakePipe(fd);
+  write(fd[1], to_write, string(to_write).length());
+  ReadPipe(fd[0], buffer_output, string(to_write).length());
+
+  EXPECT_STREQ(to_write, static_cast<const char*>(buffer_output));
+  ASSERT_DEATH(ReadPipe(-1, buffer_output, string(to_write).length()), ".*");
+  free(buffer_output);
+  ClosePipe(fd);
+}
+
+TEST_F(T_Util, ReadHalfPipe) {
+  int fd[2];
+  void *buffer_output = calloc(20, sizeof(char));
+  MakePipe(fd);
+
+  tbb::tbb_thread writer(WriteBuffer, fd[1]);
```

Comment:
Inside WriteBuffer it waits, but in the other thread.
If it wouldn't wait then the writer could write everything before the reader reads it. The idea of the ReadHalfPipe function is to read something from the pipe even if the writer hasn't started to write yet. For that there _must_ be something in the pipe, and the idea of this test was to check what happens if there is nothing written. Of course it indefinitely waits until the writer (atomically) flushes what it has written, and then the reader consumes it.